docs/enhancements/20190724-prometheus-refactoring.md

```diff
@@ -0,0 +1,370 @@
+---
+title: Refactoring of Prometheus metrics
+authors:
+  - "@aledbf"
+reviewers:
+  - "@ElvinEfendi"
+approvers:
+  - "@ElvinEfendi"
+editor: TBD
+creation-date: 2019-07-24
+last-updated: 2019-07-24
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Refactoring of Prometheus metrics
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  - [Info metrics](#info-metrics)
+    - [<em>Controller</em>](#controller)
+    - [<em>Ingresses</em> (only the leader handle this metrics)](#ingresses-only-the-leader-handle-this-metrics)
+  - [Metrics](#metrics)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Ingress-nginx currently provides prometheus metrics using Lua to extract and dispatch the data from NGINX to the ingress controller during the [log phase](http://nginx.org/en/docs/dev/development_guide.html#http_phases) making an HTTP request to a Unix socket. This socket is an HTTP server exposed by the ingress controller where the data received is used to update the defined prometheus metric. The metrics are exposed using the endpoint `/metrics` in the ingress controller.
+
+## Motivation
+
+The current implementation lacks enough configuration options to allow the customization of the features to use in the metrics. Also, due to the lack of iterations to improve metrics, the current state has several issues:
+
+- Lack of details about endpoints
+- Performance issues in scenarios with high RPS
+- Use of histogram and summaries introduce high cardinality
+- High cardinality of metrics in general.
+
+### Goals
+
+- Customizable prometheus metrics for any workload without compromising NGINX performance.
+- Metrics counter are initialized at start/reload [#4066](https://github.com/kubernetes/ingress-nginx/issues/4066)
+- Removal of flags, moving these flags to the configuration configmap. The reason for this is to not require a change in the ingress controller deployment.
+- Expose number of endpoints per ingress via prometheus [#3899](https://github.com/kubernetes/ingress-nginx/issues/3899)
+- Fix several issues:
+
+  - https://github.com/kubernetes/ingress-nginx/issues/3936
+  - https://github.com/kubernetes/ingress-nginx/issues/4026
+  - https://github.com/kubernetes/ingress-nginx/issues/3645
+  - https://github.com/kubernetes/ingress-nginx/issues/3818
+  - https://github.com/kubernetes/ingress-nginx/issues/3713
+  - https://github.com/kubernetes/ingress-nginx/pull/4139
+  - https://github.com/kubernetes/ingress-nginx/issues/2924
+  - https://github.com/kubernetes/ingress-nginx/issues/3898
+  - https://github.com/kubernetes/ingress-nginx/pull/4139
```

Comment:
An additional issue https://github.com/kubernetes/ingress-nginx/issues/3818 can be solved by incorporating a `status_class:5xx` in the metric implemented as part of this proposal